### DISCUSSION TITLE
MVVM Toolkit: ObservableValidator and ObservableRecipient together ?

### DISCUSSION THREAD

**Discussion by HispidaBytes** [CommunityToolkit/dotnet#987] [Q&A]
Hi,

I am trying to migrate my application to MVVM Toolkit as MVVM framework and during the operation a question came up: 
I wanted to know if it might be a good idea to have an "Observable" type that implements the same functionality as "ObservableRecipient" and "ObservableValidator".
I find myself in the situation where I have a ViewModel class that needs both messaging functionality between ViewModels and property validation.
Could this be a good idea or are there alternatives that I haven't considered? 

Thank you 😊

**Comment by michael-hawker** [comment]
@HispidaBytes this can be done with the new source generator support added in 7.1. @Sergio0694 was there an example of this somewhere you had yet? I know you haven't really updated the docs yet as the source generator stuff is still being worked on and considered 'in preview'.

**Reply by bukowa** [reply]
Hey, it's been a while, any chance to get updated answer?

**Comment by echoix** [comment]
There's an `[ObservableRecipient]` attribute that can be used on classes that are based on `ObservableObject` (directly or indirectly), such as `ObservableValidator`. It will skip creating constructors if the same signature already exists. There is also a conflict with `ObservableRecipient.SetProperty{T}(ref T, T, bool, string?)` and `ObservableValidator.SetProperty{T}(ref T, T, bool, string?)`, so the ObservableRecipient one(s) won't be created.

If you need to have your constructors, and also need the ones generated for ObservableRecipient, but are skipped (symptom: `Messenger` instance not initialized when leaving the constructor), create a class that is based on ObervableValidator, and has the `[ObservableRecipient]` attribute, and use that class as your base class for the class you want as ObservableRecipient and ObservableValidator. That way, you can keep a parameterless constructor for you own, and the two other parameterless constructors will still be called and used.

Pseudocode:
```c#
[ObservableRecipient]
public partial class ObservableRecipientValidator: ObservableValidator { }

// Your class
public partial class MyClass: ObservableRecipientValidator { }
```
